### ISSUE TITLE
torch.cat is 40x slower than numpy.concatenate

### ISSUE THREAD

**Issue by Staten-Wang** [pytorch/pytorch#92142]
### 🐛 Describe the bug

Due to the need, I usually need to connect close to 30k tensors, but I find that they run very slowly. I was trying to find a faster way, and I found that numpy.concatenate() is much faster than torch.cat().

I know #44025 discusses that, but there's more discussion of the number of threads and number of processes in that question.
In fact, I didn't limit the program.
When I run it, torch.cat can use all the cores of the computer (100% occupancy).

Now, I first convert all tensors to numpy and concatenate them through concatenate(). In fact, cat() and concatenate() should be about the same speed, I don't understand why there is such a huge gap between the two functions.
### To Reproduce
    import torch
    import datetime
    import time 
    import numpy as np 

    if __name__ == '__main__':
        print("strat")
        all_feature = []
        np_featyre = [] 
        for i in range(30000):
            tesnor_ = torch.rand(4,256,62)
            all_feature.append(tesnor_)
            np_featyre.append(tesnor_.numpy())
        
        start = time.time()
        all_np_feature = np.concatenate(np_featyre,0)
        print(f'{time.time() - start} s')

        start = time.time()
        all_feature = torch.cat(all_feature, dim=0)
        print(f'{time.time() - start} s')

### Output

    strat
    6.966759204864502 s
    242.34323287010193 s

### Environment
    PyTorch version: 1.11.0+cu102
    Is debug build: False
    CUDA used to build PyTorch: 10.2
    ROCM used to build PyTorch: N/A

    OS: CentOS Linux release 7.9.2009 (Core) (x86_64)
    GCC version: (GCC) 4.8.5 20150623 (Red Hat 4.8.5-44)
    Clang version: Could not collect
    CMake version: version 2.8.12.2
    Libc version: glibc-2.17

    Python version: 3.8.12 (default, Oct 12 2021, 13:49:34)  [GCC 7.5.0] (64-bit runtime)
    Python platform: Linux-3.10.0-1160.el7.x86_64-x86_64-with-glibc2.17
    Is CUDA available: True
    CUDA runtime version: 11.4.120
    CUDA_MODULE_LOADING set to:
    GPU models and configuration:
    GPU 0: Quadro RTX 8000
    GPU 1: Quadro RTX 8000
    GPU 2: Quadro RTX 8000
    GPU 3: Quadro RTX 8000

    Nvidia driver version: 470.57.02
    cuDNN version: Probably one of the following:
    /usr/local/cuda-11.4/targets/x86_64-linux/lib/libcudnn.so.8
    /usr/local/cuda-11.4/targets/x86_64-linux/lib/libcudnn_adv_infer.so.8
    /usr/local/cuda-11.4/targets/x86_64-linux/lib/libcudnn_adv_train.so.8
    /usr/local/cuda-11.4/targets/x86_64-linux/lib/libcudnn_cnn_infer.so.8
    /usr/local/cuda-11.4/targets/x86_64-linux/lib/libcudnn_cnn_train.so.8
    /usr/local/cuda-11.4/targets/x86_64-linux/lib/libcudnn_ops_infer.so.8
    /usr/local/cuda-11.4/targets/x86_64-linux/lib/libcudnn_ops_train.so.8
    HIP runtime version: N/A
    MIOpen runtime version: N/A
    Is XNNPACK available: True

    Versions of relevant libraries:
    [pip3] numpy==1.21.3
    [pip3] torch==1.11.0
    [pip3] torchaudio==0.11.0
    [pip3] torchvision==0.12.0
    [conda] numpy                     1.21.3                   pypi_0    pypi
    [conda] torch                     1.11.0                   pypi_0    pypi
    [conda] torchaudio                0.11.0                   pypi_0    pypi
    [conda] torchvision               0.12.0                   pypi_0    pypi

**Comment by mingfeima** [comment]
@Staten-Wang could you list your CPU type? number of CPU cores?

The current implementation for `cat` on CPU would do individual copies for each of the sub tensor for `input_list_length` times, which is more suitable for scenarios of short input list with large sub tensor.

For your case, it would try to parallel on the tensor shape of `[4,256,62]` for 30000 times (so the parallel session of 4x256x62 is not big enough to cover the omp parallel overhead, so to speak won't take full memory bandwidth). Note that the 30000 iteration is executed in sequential.

Anyway, I will have it fixed.

**Comment by Staten-Wang** [comment]
> @Staten-Wang could you list your CPU type? number of CPU cores?
> 
> The current implementation for `cat` on CPU would do individual copies for each of the sub tensor for `input_list_length` times, which is more suitable for scenarios of short input list with large sub tensor.
> 
> For your case, it would try to parallel on the tensor shape of `[4,256,62]` for 30000 times (so the parallel session of 4x256x62 is not big enough to cover the omp parallel overhead, so to speak won't take full memory bandwidth). Note that the 30000 iteration is executed in sequential.
> 
> Anyway, I will have it fixed.

@mingfeima my CPU is Dual Intel(R) Xeon(R) Gold 6240R CPU @ 2.40GH, 48 cores in total.
I'm a rookie, HA!
But from my observations, _concatenate()_ doesn't seem to be using parallel processing (I only see one core being 100% occupied, while _cat()_ is running on all cores.

**Comment by mingfeima** [comment]
update: replaced with https://github.com/pytorch/pytorch/pull/94373

**Comment by Staten-Wang** [comment]
Interesting!!
After many days, when I re-run this program, I get completely different results.

For numpy, I use 8.647818088531494 s
For torch, I use 2.8088386058807373 s.

And, this time both torch and numpy will call all cores. I don't understand why this happens.

**Comment by richardrl** [comment]
I'm also experiencing this. But I wonder if the reason np.concatenate looks much faster is because it plays nicer with time.perf_counter
